packages/klighd-core/src/filtering/semantic-filtering-util.ts

```diff
@@ -0,0 +1,318 @@
+ */
+export class NumericEqualConnective implements UnaryConnective {
+    static NAME = "NUMERICEQUAL"
+    name = NumericEqualConnective.NAME
+    operand: SemanticFilterTag
+    ruleName?: string
+}
+
+//// Functions ////
+
+/**
+ * A filter is used to apply a filter rule as a boolean function on a graph element. The function
+ * returns true if the element fulfils the filter rule and false otherwise.
+ */
+export interface Filter {
+    name?: string
+    defaultValue?: boolean
+    filterFun(el: SKGraphElement): boolean
+}
+
+/**
+ * Creates a new filter with a function that can be applied to graph elements when given 
+ * a filter rule.
+ * @param rule the rule to construct the filter from
+ * @returns a new filter
+ */
+export function createFilter(rule: SemanticFilterRule): Filter {
+
+    let ruleName;
+    if (rule instanceof SemanticFilterTag) {
+        ruleName = rule.tag;
+    } else {
+        ruleName = rule.ruleName;
+    }
+    return {
+        name: ruleName,
+        defaultValue: rule.defaultValue,
+        filterFun: (el) => {
+            let tags = Array<SemanticFilterTag>();
+            if (el.properties['de.cau.cs.kieler.klighd.semanticFilter.tags'] !== undefined) {
+                tags = el.properties['de.cau.cs.kieler.klighd.semanticFilter.tags'] as Array<SemanticFilterTag>;
+            }
+
+            return evaluateRule(rule, tags);
+        }
+    }
+
+}
+
+/** Evaluates `rule` using `tags`. See Connectives for further explanation on evaluation. */
+function evaluateRule(rule: SemanticFilterRule, tags: Array<SemanticFilterTag>): boolean {
+    if ((rule as SemanticFilterTag).tag !== undefined) {
+        return tags.some((tag: SemanticFilterTag) => tag.tag === (rule as SemanticFilterTag).tag);
+    }
+
+    // Rule is a Connective
+    const unary = rule as UnaryConnective;
+    const binary = rule as BinaryConnective;
+    const ternary = rule as TernaryConnective;
+    let correspondingTag;
+    switch ((rule as Connective).name) {
+        // Logic Connectives
+        case NegationConnective.NAME:
+            return !(evaluateRule(unary.operand, tags));
+        case AndConnective.NAME:
+            return evaluateRule(binary.leftOperand, tags)
+                && evaluateRule(binary.rightOperand, tags);
+        case OrConnective.NAME:
+            return evaluateRule(binary.leftOperand, tags)
+                || evaluateRule(binary.rightOperand, tags);
+        case IfThenConnective.NAME:
+            return !evaluateRule(binary.leftOperand, tags)
+                || evaluateRule(binary.rightOperand, tags);
+        case LogicEqualConnective.NAME:
+            return evaluateRule(binary.leftOperand, tags)
+                === evaluateRule(binary.rightOperand, tags);
+        case IfThenElseConnective.NAME:
+            return evaluateRule(ternary.firstOperand, tags)
+                ? evaluateRule(ternary.secondOperand, tags)
+                : evaluateRule(ternary.thirdOperand, tags);
```

Comment:
I think the evaluation of the directives would be better placed in the namespace of the directive classes in some `evaluate` function.

Reply:
could that not cause problems with the deserialization?

Reply:
serialization only uses fields and no methods of classes, I am not quite sure about the other way around. But [Actions](https://github.com/kieler/klighd-vscode/blob/main/packages/klighd-core/src/actions/actions.ts) circumvent this by putting the function into the namespace, not into some class.

Reply:
ah ok, I can do it that way